src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt

```diff
@@ -0,0 +1,279 @@
+        solver.constraints.addAll(copiedClauses)
+        countOccurrence()
+        updateSig()
+    }
+
+    private fun removeTautologies() {
+        val isClauseRemoved = MutableList(solver.constraints.size) { false }
+        for (lit in 1..solver.varsNumber) {
+            val sz1 = litOccurrence[litPos(lit)].size
+            val sz2 = litOccurrence[litPos(-lit)].size
+            if (sz1 == 0 || sz2 == 0) {
+                continue
+            }
+            var ind1 = 0
+            var ind2 = 0
+            while (ind2 < sz2) {
+                while (ind1 < sz1 && litOccurrence[litPos(lit)][ind1] < litOccurrence[litPos(-lit)][ind2]) {
+                    ind1++
+                }
+                if (ind1 == sz1) {
+                    break
+                }
+                if (litOccurrence[litPos(lit)][ind1] == litOccurrence[litPos(-lit)][ind2]) {
+                    isClauseRemoved[litOccurrence[litPos(lit)][ind1]] = true
+                }
+                ind2++
+            }
+        }
+        val deletedClauses = mutableListOf<Int>()
+        isClauseRemoved.forEachIndexed { ind, isDeleted ->
+            if (isDeleted) {
+                deletedClauses.add(ind)
+            }
+        }
+        solver.constraints.removeAll(deletedClauses.map { solver.constraints[it] })
+
+        countOccurrence()
+        updateSig()
+    }
+
+    // return position of literal in occurrence array
+    private fun litPos(lit: Int): Int {
+        return if (lit >= 0) {
+            lit
+        } else {
+            solver.varsNumber - lit
+        }
+    }
+
+    private fun countOccurrence() {
+        litOccurrence = mutableListOf()
+        //litOccurrence.clear()
+        for (ind in 1..(2 * solver.varsNumber + 1)) {
+            litOccurrence.add(mutableListOf())
+        }
+        solver.constraints.forEachIndexed { ind, clause ->
+            for (lit in clause) {
+                if (lit > 0) {
+                    litOccurrence[lit].add(ind)
+                } else {
+                    litOccurrence[solver.varsNumber - lit].add(ind)
+                }
+            }
+        }
+    }
+
+    private fun countSig(clause: Int): Long {
+        var sz = 0L
+        solver.constraints[clause].forEach { lit -> sz = sz.or(hash[litPos(lit)]) }
+        return sz
+    }
+
+    private fun Int.clauseSize() = solver.constraints[this].size
+
+    private var clauseSig = mutableListOf<Long>()
+
+    private fun updateSig() {
+        clauseSig = List(solver.constraints.size) { ind -> countSig(ind) }.toMutableList()
+    }
+
+    private fun findSubsumed(clause: Int): Set<Int> {
+        val lit = solver.constraints[clause].minByOrNull { lit -> litOccurrence[litPos(lit)].size } ?: 0 //TODO litIndex
+        return litOccurrence[litPos(lit)].filter {
+            clause != it && clause.clauseSize() <= it.clauseSize() && subset(clause, it)
+        }.toSet()
+    }
+
+    private fun subset(cl1: Int, cl2: Int): Boolean {
```

Comment:
Consider `isSubset`. I already mentioned it in another review, if I recall correctly...